Task 3.3

dudarenko-dmitry opened the pull request.

dudarenko-dmitry committed:
Create new branch TASK_3.3.
Create class Main

dudarenko-dmitry committed:
TASK_3.3 was realised according to diagram.
???

dudarenko-dmitry committed:
Name's of components were added to TASK_3.3

mrseneca commented on src/pl/senla/lecture3/task3/product/ProductImpl.java:
```diff
@@ -0,0 +1,52 @@
+    private final IProductPart secondProductPart;
+    private final IProductPart thirdProductPart;
+
+    public ProductImpl(IProductPart firstProductPart, IProductPart secondProductPart, IProductPart thirdProductPart) {
```
Let's delete this all arguments constructor and replace it with default (empty) one

mrseneca commented on src/pl/senla/lecture3/task3/product/ProductImpl.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Override
+    public void installFirstPart(IProductPart firstProductPart) {
```
And in this methods lets install parts inside class attributes, because right now, parts installed through constructor and they only writing logs. More logically will be creating empty product, and installing parts through this methods

mrseneca reviewed (commented):
Almost perfect, lets just change one thing in comments, to create real assembly line

dudarenko-dmitry committed:
Re-wrote all logic in:
- Main
- AssemblyLineImpl
- ProductImpl

Edited toString in classes which implement IProductPart and ILineStep for more convenient presentation of their instances.

dudarenko-dmitry commented:
Re-wrote almost all logic in this TASK_3.3.
I have some doubts about implementation of ProductImpl and their methods. Please, check it. Your opinion is very important for me.

mrseneca commented on src/pl/senla/lecture3/task3/building/LineStepBody.java:
```diff
@@ -0,0 +1,26 @@
+    public IProductPart buildProductPart() {
+        System.out.print("Input body's name -> ");
+        Scanner scanner = new Scanner(System.in);
+        String name = scanner.next();
```
It's unnecessary to ask user name for part in this case. Part distinguished by class

dudarenko-dmitry replied:
Ok. I tryed to make perfect application where there every method calls new stage of production and not only shows log.
But received finally strange (or even awfull) thing )))))
There was the most difficult thing in this task to understand life-cycle of different classes.

I return to previous commit and try again with your comments.

mrseneca replied:
Yeah, it's totally fine. You are learning new stuff and some incidents will happen. Don't let this small issues upset you :)

Just try to compare OOP to real live:
AssemblyLine should be in main class, it's your factory -> AssemplyLine should contain LineSteps (not product ;)), it's different factory conveyors  -> LineSteps should create your ProductParts -> Parts should be installed into Product with methods of IProduct.

It was almost ready with previous solution. But there Product was already created with installed Parts, because you have "All Argument Constructor". If you change it to installation parts with IProduct methods (setters) it will be exactly what you need

mrseneca commented on src/pl/senla/lecture3/task3/product/ProductImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+public class ProductImpl implements IProduct{
+
+    private ILineStep body;
```
ILineStep should be part of AssemblyLine, not a part of Product. In product should only be ProductParts
Let's think about real life - when you buy car there is no part of factory machines, only car parts. Same should be here

mrseneca commented on src/pl/senla/lecture3/task3/product/ProductImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public void installSecondPart(IProductPart secondProductPart) {
+        chassis = new LineStepChassis();
```
Product should not create by himself LineSteps :/
Also you pass to method some IProductPart and then in same method create new ILineStep and installing it. This is making no sense.
The part you pass should be installed in attribute not LineStep

mrseneca reviewed (commented):
Oh wow, previous solution was almost perfect and need just small fix. Now it somehow worse. 
Please read comments and ask if you can understand something.
I suggest revert this last commit and work with previous version

dudarenko-dmitry committed:
Changed attributes and methods in ProductImpl and AssemblyLineImpl.

dudarenko-dmitry commented:
Version 3 is ready :)
There are difficulties in AssemblyLine in method assembleProduct(IProduct product): I'm not sure if it is correct but I have no ideas any more.

mrseneca commented on src/pl/senla/lecture3/task3/assembly/AssemblyLineImpl.java:
```diff
@@ -0,0 +1,18 @@
+//    private ILineStep lineStepBody;
+//    private ILineStep lineStepChassis;
+//    private ILineStep lineStepEngine;
+    private IProduct car;
```
```suggestion
    private IProduct car;
    private ILineStep body = new LineStepBody();
    private ILineStep chassis = new LineStepChassis();
    private ILineStep engine = new LineStepEngine();
```

mrseneca commented on src/pl/senla/lecture3/task3/assembly/AssemblyLineImpl.java:
```diff
@@ -0,0 +1,18 @@
+    @Override
+    public IProduct assembleProduct(IProduct product) {
+        this.car = product;
+        return car;
```
```suggestion
        IProductPart firstProductPart = body.buildProductPart();
        IProductPart secondProductPart = chassis.buildProductPart();
        IProductPart thirdProductPart = engine.buildProductPart();
        System.out.println("Production of auto components was completed.\n");

        product.installFirstPart(firstProductPart);
        product.installSecondPart(secondProductPart);
        product.installThirdPart(thirdProductPart);
        
        return product;
```

dudarenko-dmitry replied:
Perfect :)
Everything turned out so easy))))

mrseneca commented on src/pl/senla/lecture3/task3/Main.java:
```diff
@@ -0,0 +1,35 @@
+
+    public static void main(String[] args) {
+        System.out.println("Production of auto components is started.");
+        ILineStep body = new LineStepBody();
```
Just delete from this and apply suggestion :)

mrseneca reviewed (commented):
Well done :)
With assembly line you did all the work in main class, it should be just moved a little bit to AssemblyLineIml and it's gonna be perfect :)

dudarenko-dmitry committed:
Final version. I hope :)
